lib/jhove_service.rb

```diff
@@ -34,25 +34,44 @@ def tech_md_output
   # @param fileset_file [Pathname,String] the pathname of the file listing which files should be processed.  If nil, process all files.
   # @return [String] Run JHOVE to characterize all content files, returning the output file path
   def run_jhove(content_dir, fileset_file=nil)
-    `#{get_jhove_command(content_dir, fileset_file)}`
-    exitcode = $?.exitstatus
-    if (exitcode != 0)
-      raise "Error when running JHOVE against #{content_dir}"
+    raise "Content #{content_dir} not found" unless File.directory? content_dir
+    if fileset_file.nil? # a simple directory gets called directly
+      exec_command(get_jhove_command(content_dir))
+    else # a filelist gets run one by one, jhove cannot do this out of the box, so we need to run jhove file by file and then assemble the results ourselves into a single XML
+      raise "File list #{fileset_file} not found" unless File.exists? fileset_file
+      files = File.new(fileset_file).readlines
+      raise "File list #{fileset_file} empty" if files.size == 0
+      combined_xml_output = ""
+      jhove_output_xml_ng = Nokogiri::XML('')
+      files.each_with_index do |filename,i| # generate jhove output for each file in a separate xml file
+        full_path_to_file = File.join(content_dir,filename.strip)
+        output_file = @target_pathname.join("jhove_output_#{i}.xml")
+        exec_command(get_jhove_command(full_path_to_file,output_file))
+        jhove_output_xml_ng = File.open(output_file) { |f| Nokogiri::XML(f) }
+        combined_xml_output += jhove_output_xml_ng.css("//repInfo").to_xml # build up an XML string with all output
+        output_file.delete
+      end
+      jhove_output_xml_ng.root.children.each {|n| n.remove} # use of the files we built up above, strip all the children to ge the root jhove node
+      jhove_output_xml_ng.root << combined_xml_output # now add the combined xml for all files
+      File.write(jhove_output, jhove_output_xml_ng.to_xml)
```

Comment:
there is apparently also a mode where jhove will store the xml in memory and return it instead of writing it to a file, but @peetucket pointed out that this approach will leave a bit of a trail for debugging, should processing unexpectedly blow up on something it encounters in the middle of the run.